### PR TITLE
Add infrastructure to capture the request body of HTTP requests

### PR DESCRIPTION
**Pull request by felixbarny**
- Add Transaction.context.request.bodyBuffer (CharBuffer)
- Add ability to decode byte[]/byte into a CharBuffer
- Adjust serialization to take the body buffer into account
- Recycling of Transaction.context.request.bodyBuffer
- -> reading is garbage free

### CODE REVIEW

**Comment by felixbarny** [comment]
Relates to #449

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java [hunk 1]
```diff
@@ -209,6 +269,14 @@ public void copyFrom(Request other) {
         this.socket.copyFrom(other.socket);
         this.url.copyFrom(other.url);
         this.cookies.copyFrom(other.cookies);
+        if (other.bodyBuffer != null) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java]: I think we have to in order to be consistent with the `capture_body` `off|errors|transactions|all` configuration option.

File: apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java [hunk 2]
```diff
@@ -758,8 +758,15 @@ private void serializeRequest(final Request request) {
             writeField("headers", request.getHeaders());
             writeField("cookies", request.getCookies());
             // only one of those can be non-empty
-            writeField("body", request.getFormUrlEncodedParameters());
-            writeField("body", request.getRawBody());
+            if (!request.getFormUrlEncodedParameters().isEmpty()) {
+                writeField("body", request.getFormUrlEncodedParameters());
+            } else if (request.getRawBody() != null) {
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/report/serialize/DslJsonSerializer.java]: No, I'll remove that

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java [hunk 3]
```diff
@@ -102,6 +127,37 @@ public Request withRawBody(String rawBody) {
         return this;
     }
 
+    /**
+     * Gets a pooled {@link CharBuffer} to record the DB statement and associates it with this instance.
```
[the remark this answers is not in the document]
Comment (felixbarny) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java]: Yes 😅

File: apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (eyalkoren) [reply apm-agent-core/src/main/java/co/elastic/apm/agent/impl/context/Request.java]: So when set to `errors` - we need to copy to the error capture and not send within the transactions, and when set to `transactions` no need to copy - here is a good place to check I guess

**Pull request merged**